modules/kafka/tests/test_kafka_topic.c

```diff
@@ -0,0 +1,182 @@
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include <syslog-ng.h>
+#include <apphook.h>
+#include "logthrdest/logthrdestdrv.h"
+#include "kafka-dest-driver.h"
+#include <librdkafka/rdkafka.h>
+#include "kafka-dest-worker.h"
+#include "kafka-dest-driver.h"
+#include "kafka-dest-worker-internal.h"
+
+#include <criterion/criterion.h>
+#include <criterion/parameterized.h>
+
+struct topic_validity_test_params
+{
+  gchar *topic_name;
+  gboolean isvalid;
+};
+
+static void
+setup(void)
+{
+  app_startup();
+}
+
+static void
+teardown(void)
+{
+  app_shutdown();
+}
+
+TestSuite(kafka_topic, .init = setup, .fini = teardown);
+
+ParameterizedTestParameters(kafka_topic, topic_validity_tests)
+{
+  static struct topic_validity_test_params params[] =
+  {
+    {"test 1", FALSE},
+    {"", FALSE},
+    {"..", FALSE},
+    {".", FALSE},
+    {" invalidname", FALSE},
+    {"aba#2", FALSE},
+    {STRING_250_LEN, FALSE},
+    {"validname123", TRUE},
+    {"valid.name123", TRUE},
+    {"valid-name123", TRUE},
+    {"valid_name123", TRUE},
+    {STRING_249_LEN, TRUE},
+  };
+
+  return cr_make_param_array(struct topic_validity_test_params, params, sizeof(params)/sizeof(params[0]));
+}
+
+ParameterizedTest(struct topic_validity_test_params *param, kafka_topic, topic_validity_tests)
+{
+  KafkaDestDriver *driver = (KafkaDestDriver *) kafka_dd_new(configuration);
+  cr_assert_eq(kafka_dd_validate_topic_name(driver, param->topic_name), param->isvalid);
+  log_pipe_unref((LogPipe *)driver);
+}
+
+Test(kafka_topic, test_resolve_template_topic_name)
+{
+  configuration = cfg_new_snippet();
+  LogDriver *driver = kafka_dd_new(configuration);
+
+  LogTemplate *topic_name = log_template_new(configuration, NULL);
+
+  log_template_compile(topic_name, "$MSG", NULL);
+  kafka_dd_set_topic(driver, topic_name);
+  kafka_dd_set_fallback_topic(driver, "fallbacktopicname");
+
+  g_assert(kafka_dd_init((LogPipe *) driver));
+
+  KafkaDestDriver *kafka_driver = (KafkaDestDriver *) driver;
+
+  KafkaDestWorker *worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
+
+  LogMessage *msg = log_msg_new_empty();
+  log_msg_set_value_by_name(msg, "MSG", "valid_template_topic", -1);
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "valid_template_topic");
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
+
+  log_msg_set_value_by_name(msg, "MSG", "invalid name", -1);
+
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "fallbacktopicname");
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
+  log_msg_unref(msg);
+
+  log_template_unref(topic_name);
+  log_threaded_dest_worker_free(&worker->super);
+  log_pipe_unref((LogPipe *)driver);
+  cfg_free(configuration);
+}
```

Comment:
I am wondering how the return value of _calculate_topic() is managed. Is it an allocated variable? Who frees that if it's allocated?

Reply:
The return type of ```_calculate_topic()``` which is a topic is freed at the end when the hashmap is freed. Incase it is non template, it is free at the end when we free the topic.
```
if (self->topic)
    rd_kafka_topic_destroy(self->topic);
```
```
  if (self->topics)
    g_hash_table_unref(self->topics);
```